src/trace.rs

```diff
@@ -0,0 +1,331 @@
+            .is_empty()
+    }
+
+    fn accumulate<T, F>(&self, condition: F) -> Vec<T>
+    where
+        F: Fn(&TraceNode) -> Option<T>,
+    {
+        let mut values = Vec::new();
+        self.walk(|n| {
+            if let Some(v) = condition(n) {
+                values.push(v)
+            }
+        });
+        values
+    }
+    fn walk<F>(&self, mut callback: F)
+    where
+        F: FnMut(&TraceNode),
+    {
+        self.walk_helper(&mut callback)
+    }
+    fn walk_helper<F>(&self, callback: &mut F)
+    where
+        F: FnMut(&TraceNode),
+    {
+        callback(self);
+        for child in &self.0.read().expect("poisoned TraceNode").children {
+            child.walk_helper(callback)
+        }
+    }
+
+    /// All the nodes descending from this node.
+    pub fn progeny(&self, states: &[TraceState]) -> Vec<TraceNode> {
+        self.accumulate(|n| {
+            if states.contains(&n.state()) {
+                Some(n.clone())
+            } else {
+                None
+            }
+        })
+    }
+    /// All the leaf nodes that descend from this node.
+    pub fn leaves(&self, states: &[TraceState]) -> Vec<TraceNode> {
+        self.accumulate(|n| {
+            if states.contains(&n.state()) && n.is_leaf() {
+                Some(n.clone())
+            } else {
+                None
+            }
+        })
+    }
+    /// Like `leaves`, but returns `Term`s instead of `TraceNodes`s.
+    pub fn leaf_terms(&self, states: &[TraceState]) -> Vec<Term> {
+        self.accumulate(|n| {
+            if states.contains(&n.state()) && n.is_leaf() {
+                Some(n.term())
+            } else {
+                None
+            }
+        })
+    }
+}
+impl PartialEq for TraceNode {
+    fn eq(&self, other: &TraceNode) -> bool {
+        Arc::ptr_eq(&self.0, &other.0)
+    }
+}
+impl Eq for TraceNode {}
+impl PartialOrd for TraceNode {
+    fn partial_cmp(&self, other: &TraceNode) -> Option<Ordering> {
+        self.log_p().partial_cmp(&other.log_p())
+    }
+}
+impl Ord for TraceNode {
+    fn cmp(&self, other: &TraceNode) -> Ordering {
+        if let Some(x) = self.partial_cmp(&other) {
+            x
+        } else {
+            Ordering::Equal
+        }
+    }
+}
+
+#[inline(always)]
+fn logsumexp(lps: &[f64]) -> f64 {
+    let largest = lps.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
+    let x = lps.iter().map(|lp| (lp - largest).exp()).sum::<f64>().ln();
+    largest + x
+}
+
+/// Samples an item from `xs` given the weights `ws`.
+fn weighted_sample<'a, T, R: Rng>(rng: &mut R, xs: &'a [T], ws: &[f64]) -> &'a T {
+    assert_eq!(xs.len(), ws.len(), "weighted sample given invalid inputs");
+    let total = ws.iter().fold(0f64, |acc, x| acc + x);
+    let threshold: f64 = Uniform::new(0f64, total).sample(rng);
+    let mut cum = 0f64;
+    for (wp, x) in ws.into_iter().zip(xs) {
+        cum += *wp;
+        if threshold <= cum {
+            return x;
+        }
+    }
+    unreachable!()
+}
```

Comment:
I think they're fine to keep here. `logsumexp` is really simple and adding a dependency for it would be overkill. Maybe something like `weighted_sample` could go into `rand`, but it would complicate their approach as they already have a weighted sampler based on relative integer weights.